local_planner/src/nodes/planner_functions.cpp

```diff
@@ -305,54 +303,24 @@ void padPolarMatrix(const Eigen::MatrixXf& matrix, unsigned int n_lines_padding,
-  Eigen::Vector3f projected_candidate = polarHistogramToCartesian(p_pol, position);
-  PolarPoint heading_pol(e_angle, azimuth_hist_frame_deg, goal_dist);
-  Eigen::Vector3f projected_heading = polarHistogramToCartesian(heading_pol, position);
-  Eigen::Vector3f projected_goal = goal;
-  PolarPoint last_wp_pol = cartesianToPolarHistogram(last_sent_waypoint, position);
-  last_wp_pol.r = goal_dist;
-  Eigen::Vector3f projected_last_wp = polarHistogramToCartesian(last_wp_pol, position);
-
-  // goal costs
+float costFunction(const PolarPoint& candidate_polar, float obstacle_distance, const Eigen::Vector3f& goal,
+                   const Eigen::Vector3f& position, const Eigen::Vector3f& velocity, const costParameters& cost_params,
+                   float& distance_cost, float& other_costs) {
+  // Compute  polar direction to goal and cartesian representation of current direction to evaluate
+  const PolarPoint facing_goal = cartesianToPolarHistogram(goal, position);
+  const Eigen::Vector3f candidate_velocity_cartesian =
+      polarHistogramToCartesian(candidate_polar, Eigen::Vector3f(0.0f, 0.0f, 0.0f));
+
+  float velocity_cost =
+      cost_params.velocity_cost_param * (velocity.norm() - candidate_velocity_cartesian.normalized().dot(velocity));
   float yaw_cost =
-      cost_params.goal_cost_param * (projected_goal.topRows<2>() - projected_candidate.topRows<2>()).norm();
-  float pitch_cost_up = 0.0f;
-  float pitch_cost_down = 0.0f;
-  if (projected_candidate.z() > projected_goal.z()) {
-    pitch_cost_up = cost_params.goal_cost_param * std::abs(projected_goal.z() - projected_candidate.z());
-  } else {
-    pitch_cost_down = cost_params.goal_cost_param * std::abs(projected_goal.z() - projected_candidate.z());
-  }
-
-  // smooth costs
-  float yaw_cost_smooth =
-      cost_params.smooth_cost_param * (projected_last_wp.topRows<2>() - projected_candidate.topRows<2>()).norm();
-  float pitch_cost_smooth = cost_params.smooth_cost_param * std::abs(projected_last_wp.z() - projected_candidate.z());
-
-  // heading cost
-  float heading_cost =
-      cost_params.heading_cost_param * (projected_heading.topRows<2>() - projected_candidate.topRows<2>()).norm();
-
-  // distance cost
-  distance_cost = 0.0f;
-  if (obstacle_distance > 0.0f) {
-    distance_cost = 700.0f / obstacle_distance;
-  }
-
-  // combine costs
-  other_costs = 0.0f;
-  other_costs = yaw_cost + cost_params.height_change_cost_param_adapted * pitch_cost_up +
-                cost_params.height_change_cost_param * pitch_cost_down + yaw_cost_smooth + pitch_cost_smooth +
-                heading_cost;
+      cost_params.yaw_cost_param * (candidate_polar.z - facing_goal.z) * (candidate_polar.z - facing_goal.z);
+  float pitch_cost =
+      cost_params.pitch_cost_param * (candidate_polar.e - facing_goal.e) * (candidate_polar.e - facing_goal.e);
+  distance_cost =
+      obstacle_distance > 0 ? 2000.0f / (std::exp((obstacle_distance - cost_params.obstacle_cost_param))) : 0.0f;
```

Comment:
I still think a reversed sigmoid function would better represent the cost here.

Eg. 100/((1 + exp(obstacle_distance - cost_params.obstacle_cost_param))^2)

Reply:
agreed, an upper bound is desirable